interrupt/interrupt.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package interrupt
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+// terminationSignals are the default signals that cause the program to exit in the
+// supported platforms (linux, darwin, windows).
+var terminationSignals = []os.Signal{syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT}
+
+// Handler guarantees execution of notifications after a critical section (the function passed
+// to a Run method), even in the presence of process termination. It guarantees exactly once
+// invocation of the provided notify functions.
+type Handler struct {
+	notify []func()
+	final  func(os.Signal)
+	once   sync.Once
+}
+
+// Chain creates a new handler that invokes all notify functions when the critical section exits
+// and then invokes the optional handler's notifications. This allows critical sections to be
+// nested without losing exactly once invocations. Notify functions can invoke any cleanup needed
+// but should not exit (which is the responsibility of the parent handler).
+func Chain(handler *Handler, notify ...func()) *Handler {
+	if handler == nil {
+		return New(nil, notify...)
+	}
+	return New(handler.signal, append(notify, handler.close)...)
```

Comment:
> Why do we pass both handler.Signal (which calls all functions and then final) and handler.Close (which calls all functions but NOT final) ? Shouldn't it be enough to do one?

The reason for that is to maintain the guarantees we state in `New` method description.

Reply:
consider:

```
h1 := New(myfinal, foo, bar)
h2 := Chain(h1, bat, qux)
h2.Run(something)
```

In the case of normal completion,  we call `h2.close()` which calls `bat()` and `qux()`, then `h1.close()` which calls `foo()` and `bar()`.  OK

In the case of a signal being delivered, we call `h2.signal()` which calls `bat()` and `qux()`, then `h1.close()` which calls `foo()` and `bar()`.  Then it calls `h1.signal()` which *DOES NOTHING* because the `once` has already been triggered.  So `myfinal() is never called`.

I think you need to pass `handler.final` not `handler.signal` ?

Additionally this does not cleanly handle sub-sections - is it supposed to ?  See trouble example at Run() below.

This whole Chain() API feels fraught to me.  Because of the `once`, calling Chain() is effectively mutating (or invalidating) the original handler.  Can we make that more obvious in the API?

Either say that Chain() "copies" the parent handler's contents (and do `append(notify, handler.notify...)`) or say that Chain() takes modifies the original handler (e.g. (`h := New(myfinal, foo, bar); h.Chain(bat, baz); h.Run()` or something.

Reply:
a fully formed example.  Note that myfinal is never called in the chain case

```
package main

import (
    "fmt"
    "os"
    "syscall"
    "time"
)

func main() {
    sig()
    chain()
}

func sig() {
    fmt.Println("TEST SIGNAL")
    fmt.Println("open foo")
    fmt.Println("open bar")
    h1 := New(myfinal, foo, bar)
    h1.Run(func() error {
        raise(syscall.SIGTERM)
        time.Sleep(100 * time.Millisecond)
        return nil 
    })  
    fmt.Println("")
}

func chain() {
    fmt.Println("TEST CHAIN SIGNAL")
    fmt.Println("open foo")
    fmt.Println("open bar")
    h1 := New(myfinal, foo, bar)
    fmt.Println("open qux")                                                                                                 
    h2 := Chain(h1, qux)
    h2.Run(func() error {
        raise(syscall.SIGTERM)
        time.Sleep(100 * time.Millisecond)
        return nil 
    })  
    fmt.Println("")
}
func myfinal(s os.Signal) {
    fmt.Printf("caught signal %q\n", s)
}
func foo() {
    fmt.Println("close foo")
}
func bar() {
    fmt.Println("close bar")
}
func qux() {
    fmt.Println("close qux")
}

func raise(sig os.Signal) error {
    p, err := os.FindProcess(os.Getpid())
    if err != nil {
        return err
    }
    return p.Signal(sig)
}                                  
```